### PR TITLE
Update to use new bitcode library structure

### PR DESCRIPTION
**Pull request by aaronenyeshi**
Also, continue support for old bitcode library layout
for backwards compatibility.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -24,7 +24,17 @@ else()
...
-    set(BITCODE_DIR "${AMD_DEVICE_LIBS_PREFIX}/lib" CACHE STRING "")
+
+    # Support building with new and old device-lib paths
+    # Previously installed to /opt/rocm/lib
+    # New layout installed to /opt/rocm/amdgcn/bitcode
+    find_path(BC_DIR NAMES "opencl.bc" "opencl.amdgcn.bc"
+      PATHS
+        "${AMD_DEVICE_LIBS_PREFIX}/amdgcn/bitcode"
+        "${AMD_DEVICE_LIBS_PREFIX}/lib/bitcode"
+        "${AMD_DEVICE_LIBS_PREFIX}/lib")
```
Comment (pfultz2) [inline CMakeLists.txt:35]: This probably needs `NO_DEFAULT_PATH` so it only searches the paths in the device libs found.

Comment (aaronenyeshi) [reply CMakeLists.txt:35]: Sure I will add that

**Comment by aaronenyeshi** [comment]
@pfultz2 , I think we may want to remove the checks for clang and bitcode dir, in the CI system, we are forcing them to the installation location, but they don't exist during build.
```
# build_clang-ocl.sh
/home/aaron/git/compute/out/ubuntu-18.04/18.04/build/clang-ocl /home/aaron/git/compute
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
CMake Error at CMakeLists.txt:42 (message):
  Clang is not found.


CMake Error at CMakeLists.txt:46 (message):
  Device libs is not found.



--------CLANG_BIN: /opt/rocm-3.7.0-9999/llvm/bin
------BITCODE_DIR: /opt/rocm-3.7.0-9999/lib

-- Triple: amdgcn-amd-amdhsa
-- Configuring incomplete, errors occurred!
See also "/home/aaron/git/compute/out/ubuntu-18.04/18.04/build/clang-ocl/CMakeFiles/CMakeOutput.log".
```

**Comment by aaronenyeshi** [comment]
During CI build, they are at 
/home/aaron/git/compute/out/ubuntu-18.04/18.04/llvm/bin/clang
/home/aaron/git/compute/out/ubuntu-18.04/18.04/amdgcn/bitcode/opencl.bc
But we want them set to this for the installed clang-ocl.
--------CLANG_BIN: /opt/rocm-3.7.0-9999/llvm/bin
------BITCODE_DIR: /opt/rocm-3.7.0-9999/amdgcn/bitcode

**Comment by aaronenyeshi** [comment]
Adding a CI_BUILD variable that our CI can set ON, so that it can skip these checks works, if you are okay with it.

**Comment by pfultz2** [comment]
> Adding a CI_BUILD variable that our CI can set ON, so that it can skip these checks works, if you are okay with it.

Yes that works. These checks are useful when the user is building it themselves from source.

Although, since we install everything into a seperate prefix(ie `/opt/rocm-3.7.0-9999`), I dont understand why the CI can't use that install prefix. Other builds will have a different version prefix so there shouldn't be a conflict.

File: CMakeLists.txt [hunk 2]
```diff
@@ -24,15 +24,29 @@ else()
...
+        "${AMD_DEVICE_LIBS_PREFIX}/lib/bitcode"
+        "${AMD_DEVICE_LIBS_PREFIX}/lib"
+      NO_DEFAULT_PATH)
+
+    set(BITCODE_DIR "${BC_DIR}" CACHE STRING "")
 endif()
 
-if(NOT EXISTS "${BITCODE_DIR}")
-    message(SEND_ERROR "Device libs is not found.")
+set(CI_BUILD OFF CACHE BOOL "Check whether this is a CI build.")
```
[the remark this answers is not in the document]
Comment (aaronenyeshi) [reply CMakeLists.txt:41]: Thanks, that's a better name for this variable.

**Pull request merged**